platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,479 @@
+     * compacted file or not.
+     */
+    private final AtomicBoolean compactionWasInProgress = new AtomicBoolean(false);
+
+    /**
+     * This variable keeps track of the compaction level that was in progress at the time when it was suspended.
+     * Once the compaction is resumed, this level is used to start a new compacted file, and then it's reset to 0.
+     */
+    private final AtomicInteger compactionLevelInProgress = new AtomicInteger(0);
+
+    /** When was the last medium-sized merge, only touched from single merge thread. */
+    private Instant lastMediumMerge;
+
+    /** When was the last full merge, only touched from single merge thread. */
+    private Instant lastFullMerge;
+
+    /** A nanosecond-precise Clock */
+    private final NanoClock clock = new NanoClock();
+
+    /** The data file collection to compact */
+    private final DataFileCollection<?> dataFileCollection;
+
+    public DataFileCompactor(DataFileCollection<?> dataFileCollection) {
+        this.dataFileCollection = dataFileCollection;
+        this.storeName = dataFileCollection.getStoreName();
+        // Compute initial merge periods to a randomized value of now +/- 50% of merge period. So
+        // each node will do
+        // medium and full merges at random times.
+        lastMediumMerge = Instant.now()
+                .minus(config.mediumMergePeriod() / 2, config.mergePeriodUnit())
+                .plus((long) (config.mediumMergePeriod() * Math.random()), config.mergePeriodUnit());
+        lastFullMerge = Instant.now()
+                .minus(config.fullMergePeriod() / 2, config.mergePeriodUnit())
+                .plus((long) (config.fullMergePeriod() * Math.random()), config.mergePeriodUnit());
+    }
+
+    /**
+     * Compacts all files in filesToCompact.
+     *
+     * @param index          takes a map of moves from old location to new location. Once it is finished and
+     *                       returns it is assumed all readers will no longer be looking in old location, so old files
+     *                       can be safely deleted.
+     * @param filesToCompact   list of files to compact
+     * @return list of files created during the compaction
+     * @throws IOException          If there was a problem with the compaction
+     * @throws InterruptedException If the compaction thread was interrupted
+     */
+    // visible for testing
+    synchronized List<Path> compactFiles(final CASableLongIndex index, final List<DataFileReader<?>> filesToCompact)
+            throws IOException, InterruptedException {
+        if (filesToCompact.size() < 2) {
+            // nothing to do we have merged since the last data update
+            logger.debug(MERKLE_DB.getMarker(), "No files were available for merging [{}]", storeName);
+            return Collections.emptyList();
+        }
+
+        int currentLevel = filesToCompact.get(0).getMetadata().getCompactionLevel();
```

Comment:
This is implicit dependency on the order of files in the list. Would it be better to pass the level as a separate arg to the method?

Reply:
As of now, all the files in `filesToCompact` have the same compaction level. So, this particular approach doesn't introduce the implicit dependency you're talking about. However, later this will change because I'll have to prepare this list of files to compact which may include files of different levels. Once I have that, I'll add logic traversing these files to find the max level.

Reply:
A full compaction includes all files of levels 0 to 3. If the first file is of level 0, the full compaction produces a level 1 file to be included in next medium compaction.